src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs

```diff
@@ -122,6 +122,9 @@ public static unsafe ReadOnlySpan<byte> AsBytes<T>(ReadOnlySpan<T> span)
             if (RuntimeHelpers.IsReferenceOrContainsReferences<TTo>())
                 ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(TTo));
 
+            if (typeof(TFrom) == typeof(TTo))
+                return new Span<TTo>(ref Unsafe.As<TFrom, TTo>(ref span._reference), span.Length);
```

Comment:
@EgorBo, without this, I'm seeing the JIT doesn't seem to quite be able to eliminate the work related to the call when the types are equal. Do you see the same thing?

Reply:
@stephentoub which types should I use to reproduce this?

Reply:
e.g. char

Reply:
I should have shared more info, sorry. The difference isn't huge, and some of it is dead code that not eliminated, but wanted to call it out.
[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4a9Jq049+ggJI8ZEMfKhKV6mpoDMTUgyEMaAbxoNPTe2xjYAJgDyXJIAnrJi2FwAPGAAFthQAHwMALLkABS+AcFhEVGxCckMuOQAlAwAvCmpMPjQoamJuAmSwngYhYloDPGJSRmlZRrUXt4M2UEh4ZExfcWppFl+U3mzXcVDVTWhQh0bPfMDQyNjHl5iULwK2Bgw2kgTK7kzBQAqbxA7e7idbwBiUFEPQ+X2WOWm+RiAKB+BSuFmFWq5zGnlq9SgjWarWEsFuMEmLyhGVgADMGABVLi4bCkwQAQVw0RhwIYoIGZLSdQaTSgLWwbQA4jAML46bAuCpBoiyj0EVEWAAZGBcADmGDiwxoAF8gA==)

Reply:
Right. so M1 comes with a frame setup (all non-leaf or comple methods will have it anway) and an unreachable `throw OverflowException` block. The latter is a known issue we'll eventually fix, but it shouldn't really affect code performance/density.

M3 looks more efficient because it realizes that it doesn't need to promote Span to registers and can copy the whole thing using just SIMD - something that we can't do if we decide to promote a struct (and it does need to be promoted for Cast<>).
So overall I would not bother special casing this.

cc @dotnet/jit-contrib

Reply:
> The SIMD copy can hit a stall

Isn't that a general problem we should be addressing? If we don't believe the SIMD copy is appropriate, why do we use it?

Reply:
Yes, we should investigate what we can do about these kinds of stalls. The JIT today does not pay much attention to what may be data and what may be padding when generating code around structs -- everything is considered to be data. Unfortunately I don't think there's always a simple optimal strategy, SIMD copies are better in some cases, field-by-field copies are better in other cases, and it takes non-local reasoning to know what's what.

One thing that may help would also be to try to increase the size of stores to also fill nearby padding when that can be done cheaply, so that we e.g. write the full last 8 bytes of `Span<T>` when `Span<T>._length` is assigned. That would probably help on these stalls too.

Coincidentally I was investigating #96524 earlier, and I'm pretty sure it's the same problem.

Reply:
Thanks. If we don't have an issue for it, let's open one.

Reply:
Good idea, I opened #100769